rustls/src/msgs/deframer/buffers.rs

```diff
@@ -0,0 +1,115 @@
+use core::mem;
+use core::ops::Range;
+
+/// Conversion from a slice within a larger buffer into
+/// a `Range` offset within.
+#[derive(Debug)]
+pub(crate) struct Locator {
+    bounds: Range<*const u8>,
+}
+
+impl Locator {
+    pub(crate) fn new(slice: &[u8]) -> Self {
+        Self {
+            bounds: slice.as_ptr_range(),
+        }
+    }
+
+    pub(crate) fn locate(&self, slice: &[u8]) -> Range<usize> {
+        let bounds = slice.as_ptr_range();
+        debug_assert!(self.fully_contains(slice));
+        let start = bounds.start as usize - self.bounds.start as usize;
+        let len = bounds.end as usize - bounds.start as usize;
+        Range {
+            start,
+            end: start + len,
+        }
+    }
+
+    pub(crate) fn fully_contains(&self, slice: &[u8]) -> bool {
+        let bounds = slice.as_ptr_range();
+        bounds.start >= self.bounds.start && bounds.end <= self.bounds.end
+    }
+}
+
+/// Conversion from a `Range` offset to the original slice.
+pub(crate) struct Delocator<'b> {
+    slice: &'b [u8],
+}
+
+impl<'b> Delocator<'b> {
+    pub(crate) fn new(slice: &'b [u8]) -> Self {
+        Self { slice }
+    }
+
+    pub(crate) fn slice_from_range(&'_ self, range: &Range<usize>) -> &'b [u8] {
+        self.slice.get(range.clone()).unwrap()
+    }
+
+    pub(crate) fn locator(self) -> Locator {
+        Locator::new(self.slice)
+    }
+}
+
+/// Reordering the underlying buffer based on ranges.
+pub(crate) struct Coalescer<'b> {
+    slice: &'b mut [u8],
+}
+
+impl<'b> Coalescer<'b> {
+    pub(crate) fn new(slice: &'b mut [u8]) -> Self {
+        Self { slice }
+    }
+
+    pub(crate) fn copy_within(&mut self, from: Range<usize>, to: Range<usize>) {
+        debug_assert!(from.len() == to.len());
+        self.slice.copy_within(from, to.start);
```

Comment:
`copy_within` says:
> This function will panic if either range exceeds the end of the slice, or if the end of src is before the start.

Is that latter condition something that should have a `debug_assert!()`?

Reply:
Added some assertions using `self.slice.get` which have the same preconditions.